### ISSUE TITLE
Block storage using erasure coding doesn't work

### ISSUE THREAD

**Issue by benceszikora** [rook/rook#1733]
**Bug Report**

What happened: When trying to use block storage with an erasure coded pool I get the following errors in the operator:
```2018-05-22 14:34:07.303570 I | op-provisioner: creating volume with configuration {pool:ecpool clusterNamespace:rook fstype:}                                                                                                                                                                         2018-05-22 14:34:07.303598 I | exec: Running command: rbd create ecpool/pvc-3ff8755a-5dcc-11e8-876a-ecf4bbc33eac --size 1024 --cluster=rook --conf=/var/lib/rook/rook/rook.config --keyring=/var/lib/rook/rook/client.admin.keyring                                                                   
E0522 14:34:10.101498       7 controller.go:801] Failed to provision volume for claim "default/test-ec-claim" with StorageClass "rook-block": Failed to create rook block image ecpool/pvc-3ff8755a-5dcc-11e8-876a-ecf4bbc33eac: failed to create image pvc-3ff8755a-5dcc-11e8-876a-ecf4bbc33eac in po
ol ecpool of size 1073741824: Failed to complete '': exit status 95. 2018-05-22 14:34:10.081449 7fffd3fff700 -1 librbd::image::CreateRequest: 0x555555b6cfc0 handle_add_image_to_directory: error adding image to directory: (95) Operation not supported                                             
````

What you expected to happen: The volume is created and attached.

How to reproduce it (minimal and precise):
Apply:
```apiVersion: rook.io/v1alpha1
kind: Pool
metadata:
  name: ecpool
  namespace: rook
spec:
  failureDomain: osd
  erasureCoded:
    dataChunks: 8
    codingChunks: 2
---
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
   name: rook-block
provisioner: rook.io/block
parameters:
  pool: ecpool
````
And create a pod with a PVC using rook-block.

From what I understand based on [https://www.spinics.net/lists/ceph-users/msg39489.html](this) , the issue is that the metadata still needs to use a replicated pool. Following the suggestions in that thread, that does indeed seem to work:

```root@rook-tools:/# rbd create ecpool/test --size 10                      
2018-05-22 14:40:24.193248 7fffd3fff700 -1 librbd::image::CreateRequest: 0x555555b64250 handle_add_image_to_directory: error adding image to directory: (95) Operation not supported
rbd: create error: (95) Operation not supported                          
root@rook-tools:/# rbd create --size 10 --data-pool ecpool replicapool/test                                                                        
root@rook-tools:/# rbd create ecpool/test --size 10                                                                                                
2018-05-22 14:42:06.132323 7fffd3fff700 -1 librbd::image::CreateRequest: 0x555555b63fd0 handle_add_image_to_directory: error adding image to directory: (95) Operation not supported
rbd: create error: (95) Operation not supported                          
root@rook-tools:/# rbd info replicapool/test                             
rbd image 'test':                   
        size 10240 kB in 3 objects  
        order 22 (4096 kB objects)  
        data_pool: ecpool           
        block_name_prefix: rbd_data.4.1ec7e2ae8944a                      
        format: 2                   
        features: layering, exclusive-lock, object-map, fast-diff, deep-flatten, data-pool                                                         
        flags:                      
        create_timestamp: Tue May 22 14:41:42 2018
```

**Environment**:
* OS (e.g. from /etc/os-release): RHEL 7.3
* Kernel (e.g. `uname -a`): 3.10.0-514.6.1.el7
* Rook version (use `rook version` inside of a Rook Pod): v0.7.0-100.ge11b3d8
* Kubernetes version (use `kubectl version`): v1.9.3
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): Open source on-prem
* Ceph status: HEALTH_OK

**Comment by benceszikora** [comment]
@tonya11en I was using an older version that I thought, I'll give it a go on master now, thanks!

**Comment by benceszikora** [comment]
@tonya11en I have tested this again on v0.7.0-126.gf6abaec and I am still seeing the same issue. Is there a something I am not setting, or should it just work with this config?

**Comment by tonya11en** [comment]
Using your provided config on an even older Rook version (rook: v0.7.0-86.g6d7ca2c9), I'm creating the following:
```
>> kc -n rook describe pool/ecpool
Name:         ecpool
Namespace:    rook
Labels:       <none>
Annotations:  <none>
API Version:  rook.io/v1alpha1
Kind:         Pool
Metadata:
  Cluster Name:        
  Creation Timestamp:  2018-05-29T17:41:48Z
  Generation:          1
  Resource Version:    1225
  Self Link:           /apis/rook.io/v1alpha1/namespaces/rook/pools/ecpool
  UID:                 9044caf3-6367-11e8-a773-08002750cadf
Spec:
  Erasure Coded:
    Coding Chunks:  2
    Data Chunks:    8
  Failure Domain:   osd
Events:             <none>
```

If the Rook cluster is using the default `storeType`, then I get the following error on creation:
```
2018-05-29 17:42:19.120656 I | op-pool: creating pool ecpool in namespace rook
2018-05-29 17:42:19.121236 I | exec: Running command: ceph osd erasure-code-profile get default --cluster=rook --conf=/var/lib/rook/rook/rook.config --keyring=/var/lib/rook/rook/client.admin.keyring --format json --out-file /tmp/654426999
2018-05-29 17:42:19.310330 I | exec: Running command: ceph osd erasure-code-profile set ecpool_ecprofile k=8 m=2 plugin=jerasure technique=reed_sol_van crush-failure-domain=osd --cluster=rook --conf=/var/lib/rook/rook/rook.config --keyring=/var/lib/rook/rook/client.admin.keyring --format json --out-file /tmp/033966698
2018-05-29 17:42:20.534924 I | exec: Running command: ceph osd pool create ecpool 0 erasure ecpool_ecprofile --cluster=rook --conf=/var/lib/rook/rook/rook.config --keyring=/var/lib/rook/rook/client.admin.keyring --format json --out-file /tmp/482408129
2018-05-29 17:42:22.595638 I | exec: pool 'ecpool' created
2018-05-29 17:42:22.596526 I | exec: Running command: ceph osd pool set ecpool allow_ec_overwrites true --cluster=rook --conf=/var/lib/rook/rook/rook.config --keyring=/var/lib/rook/rook/client.admin.keyring --format json --out-file /tmp/909102380
2018-05-29 17:42:22.858124 I | exec: Error EINVAL: pool must only be stored on bluestore for scrubbing to work: osd.0 uses filestore
2018-05-29 17:42:22.858395 E | op-pool: failed to create pool ecpool. failed to create pool ecpool. failed to allow EC overwrite for pool ecpool. failed to set pool property allow_ec_overwrites on pool ecpool, exit status 22
```

However, setting `storeType: bluestore` yields no error. Are you using bluestore or the default storeType?

**Comment by benceszikora** [comment]
We are using bluestore, this is the cluster config I am using:
```kind: Cluster                       
metadata:                           
  name: rook                        
  namespace: rook                   
spec:                               
  dataDirHostPath: /data/rook/      
  storage:                          
    useAllNodes: true               
    useAllDevices: false            
    storeConfig:                    
      storeType: bluestore          
    metadataDevice: "/dev/nvme0n1"  
    deviceFilter: "^((sd[e-z])|(sda[a-s]))"                              
```

The error is the same even after updating:
```
. output: 2018-05-30 08:55:25.192190 7fffd3fff700 -1 librbd::image::CreateRequest: 0x555555b71d20 handle_add_image_to_directory: error adding image to directory: (95) Operation not supported
E0530 08:55:25.212401       7 goroutinemap.go:165] Operation for "provision-default/ec-test-claim-2[2ce31f84-63e7-11e8-9d71-ecf4bbc33eac]" failed. No retries permitted until 2018-05-30 08:55:26.212385308 +0000 UTC m=+68164.505609967 (durationBeforeRetry 1s). Error: Failed to create rook block image ecpool/pvc-2ce31f84-63e7-11e8-9d71-ecf4bbc33eac: failed to create image pvc-2ce31f84-63e7-11e8-9d71-ecf4bbc33eac in pool ecpool of size 1073741824: Failed to complete '': exit status 95. 2018-05-30 08:55:25.192190 7fffd3fff700 -1 librbd::image::CreateRequest: 0x555555b71d20 handle_add_image_to_directory: error adding image to directory: (95) Operation not supported
```

**Comment by galexrt** [comment]
@tonya11en The PR you point does not fix this completely.

From what I read about erasure coded (at least for) RBD pools, two pools are needed, kind of like with CephFS having two pools (by default).
One for metadata and for data, so when `rbd create` is called the `--data-pool` option needs to be given pointing to the data pool but for the "image prefix" one needs to use the metadata pool[1].

[1] https://ceph.com/community/new-luminous-erasure-coding-rbd-cephfs/ section "Using EC pools with RBD"

**Comment by AntonioMeireles** [comment]
@galexrt thanks for clarifying this!

**Comment by jbw976** [comment]
added as nice to have in v0.8 since this was intended to be fixed entirely in #1129 via #1651, but there is apparently more work to do.

**Comment by AntonioMeireles** [comment]
@jbw976 in the meantime (specially *if* getting it  merged ends taking a fair bit) IMHO it may make sense to clearly indicate in the `master` docs that as of now this is not working, to manage users's expectations ...

**Comment by jbw976** [comment]
as per @galexrt, we should look also at the docs/samples to see if the default paths are using erasure coding.  if EC doesn't work by default for block storage, that should not be the default.

**Comment by liewegas** [comment]
even if it does work, i'm not sure EC is a good default for RBD given the performance tradeoff.

**Comment by galexrt** [comment]
> Is it a bird? Is it a plane? No, it's a sunny weekend.

I had some time this weekend to create a PR that can "fix" this, see #1959.

**Comment by colinkelly** [comment]
Why is this closed? We have now metadatapool and datapool but same issue.

Can somebody correct me if I'm wrong?
